### ISSUE TITLE
hierarchical definition of domains

### ISSUE THREAD

**Issue by aldopareja** [dragonfly/dragonfly#63]
how could I define non-parametric or hierarchical domains?.

A hierarchical example, you can choose between SGD and ADAM, if you choose the first one, you would have to specify learning rate and momentum. If you choose the second one you would have to specify gamma and beta on top of the learning rate. There's also clipping values if you have some clipping optimizer.

As a non-parametric example, one could choose to add a new layer, which could be a fully connected, batch_norm or dropout. Depending on the type of layer there are several parameters to be specified (eg. dropout rate or number of units). Thing is that the number of layers is potentially infinite.

Do you think I should opt for different optmization algos?. Something like evolutionary algos or something else?
